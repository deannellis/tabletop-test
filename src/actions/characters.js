import uuid from 'uuid';

// ADD_CHARACTER
export const addCharacter = (
    {
        abilities = {
            strength: 0,
            intelligence: 0,
            wisdom: 0,
            dexterity: 0,
            constitution: 0,
            charisma: 0
        }, 
        race = '', 
        charClass = '', 
        hp = 0, 
        xp = 0,
        name = '',
        description = ''
    } = {}
) => ({
    type: 'ADD_CHARACTER',
    character: {
        id: uuid(),
        abilities,
        race,
        charClass,
        hp,
        xp,
        name,
        description
    }
});

// REMOVE_CHARACTER

export const removeCharacter = ({ id } = {}) => ({
    type: 'REMOVE_CHARACTER',
    id
});

// EDIT_CHARACTER
export const editCharacter = (id, updates) => ({
    type: 'EDIT_CHARACTER',
    id,
    updates
});

