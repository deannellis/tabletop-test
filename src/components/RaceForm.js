import React from 'react';

export default class AbilitiesForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            race: undefined
        };
    }
    
    render() {
        return (
            <div>
                <h2>Select Your Race</h2>
            </div>
        )
    }
}