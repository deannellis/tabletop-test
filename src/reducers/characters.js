//  CHARACTERS REDUCER
import actionIds from 'actions/actionIds';
const {ADD_CHARACTER, REMOVE_CHARACTER, EDIT_CHARACTER, COMPLETE_CHARACTER} = actionIds;
const charactersReducerDefaultState = [];

export default ( state = charactersReducerDefaultState, action ) => {
    switch (action.type) {
        case ADD_CHARACTER:
            return [
                ...state,
                action.character
            ];
        case REMOVE_CHARACTER:
            return state.filter(({ id }) => {
                return id !== action.id;
            });
        case EDIT_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    return {
                        ...character,
                        ...action.updates
                    }
                } else {
                    return character;
                }
            });
        case COMPLETE_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    let {inProgressStep, ...rest} = character;
                    console.log('rest'+rest);
                    return rest;
                } else {
                    return character;
                }
            });
        default:
            return state;
    }
};