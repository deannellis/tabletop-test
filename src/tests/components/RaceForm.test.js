/* global jest expect */
import React from 'react';
import { shallow } from 'enzyme';
import RaceForm from '../../components/RaceForm';
import characters from '../fixtures/characters';

test('should render race form without props (character not found)', () => {
    const wrapper = shallow(<RaceForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render race form with props (all options enabled)', () => {
    const wrapper = shallow(<RaceForm currentCharacter={characters[0]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render race form with low constitution (dwarf disabled)', () => {
    const wrapper = shallow(<RaceForm currentCharacter={characters[3]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render race form with low intelligence (elf disabled)', () => {
    const wrapper = shallow(<RaceForm currentCharacter={characters[5]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render race form with low dexterity (halfling disabled)', () => {
    const wrapper = shallow(<RaceForm currentCharacter={characters[6]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid submission', () => {
    const wrapper = shallow(<RaceForm currentCharacter={characters[1]}/>);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test('should set selected option on change', () => {
    const wrapper = shallow(<RaceForm currentCharacter={characters[1]}/>);
    const value = "dwarf";
    wrapper.find('input').at(0).simulate('change', {
       target: { value }
    });
    expect(wrapper.state('selectedOption')).toBe(value);
});

test('should call onSubmit prop for valid submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<RaceForm currentCharacter={characters[1]} onSubmit={onSubmitSpy} />);
    const value = "dwarf";
    wrapper.find('input').at(0).simulate('change', {
       target: { value }
    });
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith(
        characters[1].id,
        {
            race: value,
            inProgressStep: 3
        }
    );
});