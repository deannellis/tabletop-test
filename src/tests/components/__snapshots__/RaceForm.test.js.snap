// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render error for invalid submission 1`] = `
<div>
  <h2>
    Select Your Race
  </h2>
  <p>
    Please select a race
  </p>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="dwarf"
        />
        Dwarf
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="elf"
        />
        Elf
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="halfling"
        />
        Halfling
      </label>
    </div>
    <label>
      <input
        checked={false}
        name="race"
        onChange={[Function]}
        type="radio"
        value="human"
      />
      Human
    </label>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render race form with low constitution (dwarf disabled) 1`] = `
<div>
  <h2>
    Select Your Race
  </h2>
  <form
    onSubmit={[Function]}
  >
    <p>
      Dwarf: Your ability scores are insufficient
    </p>
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="elf"
        />
        Elf
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="halfling"
        />
        Halfling
      </label>
    </div>
    <label>
      <input
        checked={false}
        name="race"
        onChange={[Function]}
        type="radio"
        value="human"
      />
      Human
    </label>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render race form with low dexterity (halfling disabled) 1`] = `
<div>
  <h2>
    Select Your Race
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="dwarf"
        />
        Dwarf
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="elf"
        />
        Elf
      </label>
    </div>
    <p>
      Halfling: Your ability scores are insufficient
    </p>
    <label>
      <input
        checked={false}
        name="race"
        onChange={[Function]}
        type="radio"
        value="human"
      />
      Human
    </label>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render race form with low intelligence (elf disabled) 1`] = `
<div>
  <h2>
    Select Your Race
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="dwarf"
        />
        Dwarf
      </label>
    </div>
    <p>
      Elf: Your ability scores are insufficient
    </p>
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="halfling"
        />
        Halfling
      </label>
    </div>
    <label>
      <input
        checked={false}
        name="race"
        onChange={[Function]}
        type="radio"
        value="human"
      />
      Human
    </label>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render race form with props (all options enabled) 1`] = `
<div>
  <h2>
    Select Your Race
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="dwarf"
        />
        Dwarf
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="elf"
        />
        Elf
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="race"
          onChange={[Function]}
          type="radio"
          value="halfling"
        />
        Halfling
      </label>
    </div>
    <label>
      <input
        checked={false}
        name="race"
        onChange={[Function]}
        type="radio"
        value="human"
      />
      Human
    </label>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render race form without props (character not found) 1`] = `<NotFoundPage />`;
